### PR TITLE
Read hdf does not close store

### LINKED ISSUE #25766: read_hdf does not close store when raise KeyError
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
df = pd.DataFrame({'a': range(10), 'b': range(10)})

df.to_hdf('/tmp/1.hdf', 'k1')
try:
    pd.read_hdf('/tmp/1.hdf', 'k2')
except Exception as e:
    print(type(e))
    print(e)
    
df.to_hdf('/tmp/1.hdf', 'k2')
```
#### Problem description
<type 'exceptions.KeyError'>
'No object named k2 in the file'

The file '/tmp/1.hdf' is already opened, but in read-only mode.  Please close it before reopening in append mode.

#### Expected Output
should not raise Exception when: df.to_hdf('/tmp/1.hdf', 'k2')

#### pandas/io/pytables.py line 408
``` python

 except (ValueError, TypeError):  # should also catch KeyError 
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.12.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-141-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: zh_CN.UTF-8
LOCALE: None.None

pandas: 0.24.2
pytest: 3.2.4
pip: 19.0.1
setuptools: 39.1.0
Cython: 0.27.3
numpy: 1.14.5
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 5.5.0
sphinx: 1.6.5
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.1.0
openpyxl: 2.4.9
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml.etree: 4.1.1
bs4: 4.6.0
html5lib: 0.9999999
sqlalchemy: 1.1.15
pymysql: None
psycopg2: 2.7.3.2 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by rbenes**
- [x] closes #25766
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/io/test_pytables.py [hunk 1]
```diff
@@ -1223,6 +1223,16 @@ def test_append_all_nans(self):
             reloaded = read_hdf(path, 'df_with_missing')
             tm.assert_frame_equal(df_with_missing, reloaded)
 
+    def test_read_missing_key_close_store(self):
+        # GH 25766
+        with ensure_clean_path(self.path) as path:
+            df = pd.DataFrame({'a': range(2), 'b': range(2)})
+            df.to_hdf(path, 'k1')
+
+            pytest.raises(KeyError, pd.read_hdf, path, 'k2')
```
Comment (jreback) [inline pandas/tests/io/test_pytables.py]: this is not an accepted format any longer, use the context manager

```
with pytest.raises(....):
    ...
```

File: pandas/tests/io/test_pytables.py [hunk 2]
```diff
@@ -1223,6 +1223,16 @@ def test_append_all_nans(self):
...
+
```
Comment (jreback) [inline pandas/tests/io/test_pytables.py]: add a comment that this is smoke test that the file is close before written here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/io/test_pytables.py (hunk 1 above)
Comment (rbenes) [reply pandas/tests/io/test_pytables.py]: updated https://github.com/pandas-dev/pandas/pull/25863/commits/244cab4ecb1d27564580a5702a93f58235075173 . But with form `pytest.raises(KeyError, ...)` I followed the codestyle in this test file. What about some unification of this style across test files in future?

File: pandas/tests/io/test_pytables.py (hunk 2 above)
Comment (rbenes) [reply pandas/tests/io/test_pytables.py]: updated https://github.com/pandas-dev/pandas/pull/25863/commits/a53b0396504a4c448a5c779dfb5a871cab4ae080

File: pandas/tests/io/test_pytables.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/io/test_pytables.py]: yep this happens to be the last file - i think it’s fixed in another PR

File: doc/source/whatsnew/v0.25.0.rst [hunk 3]
```diff
@@ -329,6 +329,7 @@ I/O
 - Bug in :meth:`DataFrame.to_string` and :meth:`DataFrame.to_latex` that would lead to incorrect output when the ``header`` keyword is used (:issue:`16718`)
 - Bug in :func:`read_csv` not properly interpreting the UTF8 encoded filenames on Windows on Python 3.6+ (:issue:`15086`)
 - Improved performance in :meth:`pandas.read_stata` and :class:`pandas.io.stata.StataReader` when converting columns that have missing values (:issue:`25772`)
+- Bug in :func:`read_hdf` not properly close store when key is not found (:issue:`25766`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.25.0.rst]:
```suggestion
- Bug in :func:`read_hdf` not properly closing store after a KeyError is raised (:issue:`25766`)
```

**Review by WillAyd — CHANGES_REQUESTED** [review]
Minor suggestion to improve whatsnew but otherwise lgtm

**Review by WillAyd — APPROVED** [review]
lgtm @jreback

File: doc/source/whatsnew/v0.25.0.rst [hunk 4]
```diff
@@ -329,6 +329,7 @@ I/O
 - Bug in :meth:`DataFrame.to_string` and :meth:`DataFrame.to_latex` that would lead to incorrect output when the ``header`` keyword is used (:issue:`16718`)
 - Bug in :func:`read_csv` not properly interpreting the UTF8 encoded filenames on Windows on Python 3.6+ (:issue:`15086`)
 - Improved performance in :meth:`pandas.read_stata` and :class:`pandas.io.stata.StataReader` when converting columns that have missing values (:issue:`25772`)
+- Bug in :func:`read_hdf` not properly closing store after a KeyError is raised (:issue:`25766`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: use double backticks around ``KeyError``

**Review by jreback — CHANGES_REQUESTED** [review]
trivial doc-update. ping on green.

File: doc/source/whatsnew/v0.25.0.rst (hunk 4 above)
Comment (WillAyd) [reply doc/source/whatsnew/v0.25.0.rst]: Looks like this was done - going to merge

**Comment by WillAyd** [comment]
Thanks @rbenes !

**Pull request merged**